From e62e6c2b52449be276fc0885178704e4cb8be36c Mon Sep 17 00:00:00 2001
From: Tomas Henzl <thenzl@redhat.com>
Date: Thu, 16 Oct 2014 16:02:13 -0400
Subject: [scsi] mpt2sas: fix undefined reference to `__udivdi3' compilation errors

Message-id: <1413475335-14737-31-git-send-email-thenzl@redhat.com>
Patchwork-id: 97611
O-Subject: [RHEL7.1 e-stor PATCH 30/32] mpt2sas: fix undefined reference to `__udivdi3' compilation errors
Bugzilla: 1088524
RH-Acked-by: Maurizio Lombardi <mlombard@redhat.com>
RH-Acked-by: Rob Evers <revers@redhat.com>

This patch will fix the below compilation errors on i386 ARCH

drivers/built-in.o: In function `_scsih_qcmd':
mpt2sas_scsih.c:(.text+0x1e7b56): undefined reference to `__udivdi3'
mpt2sas_scsih.c:(.text+0x1e7b8a): undefined reference to `__umoddi3'

Used sector_div() API to fix above compilation errors.

Signed-off-by: Sreekanth Reddy <Sreekanth.Reddy@avagotech.com>
Reviewed-by: Martin K. Petersen <martin.petersen@oracle.com>
Signed-off-by: Christoph Hellwig <hch@lst.de>
(cherry picked from commit f78f694c34ae6fdc29f5e80abb2dbb894f961772)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/mpt2sas/mpt2sas_scsih.c b/drivers/scsi/mpt2sas/mpt2sas_scsih.c
index 992a224..c80ed04 100644
--- a/drivers/scsi/mpt2sas/mpt2sas_scsih.c
+++ b/drivers/scsi/mpt2sas/mpt2sas_scsih.c
@@ -3860,7 +3860,7 @@ _scsih_setup_direct_io(struct MPT2SAS_ADAPTER *ioc, struct scsi_cmnd *scmd,
  struct _raid_device *raid_device, Mpi2SCSIIORequest_t *mpi_request,
  u16 smid)
 {
- sector_t v_lba, p_lba, stripe_off, stripe_unit, column, io_size;
+ sector_t v_lba, p_lba, stripe_off, column, io_size;
  u32 stripe_sz, stripe_exp;
  u8 num_pds, cmd = scmd->cmnd[0];
 
@@ -3888,9 +3888,9 @@ _scsih_setup_direct_io(struct MPT2SAS_ADAPTER *ioc, struct scsi_cmnd *scmd,
 
  num_pds = raid_device->num_pds;
  p_lba = v_lba >> stripe_exp;
- stripe_unit = p_lba / num_pds;
- column = p_lba % num_pds;
- p_lba = (stripe_unit << stripe_exp) + stripe_off;
+ column = sector_div(p_lba, num_pds);
+ p_lba = (p_lba << stripe_exp) + stripe_off;
+
  mpi_request->DevHandle = cpu_to_le16(raid_device->pd_handle[column]);
 
  if (cmd == READ_10 || cmd == WRITE_10)
-- 
1.7.1